File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    Map<Node, Node> map = new HashMap<>();
    public Node cloneGraph(Node node) {
        if(node == null){
            return null;
        }
        if(map.containsKey(node)){
            return map.get(node);
        }
        map.put(node, new Node(node.val, new ArrayList<>()));
        for(int i = 0; i < node.neighbors.size(); i++){
            Node temp = cloneGraph(node.neighbors.get(i));
            map.get(node).neighbors.add(temp);
        }
        return map.get(node);
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        hashmap = {}
        def helper(node):
            if not node:
                return None
            if node in hashmap:
                return hashmap[node]
            hashmap[node] = Node(node.val)
            neighbor = node.neighbors
            for i in range(len(neighbor)):
                hashmap[node].neighbors.append(helper(neighbor[i]))
            return hashmap[node]
        return helper(node)
